pkg/controller/storagecluster/storagecluster.go

```diff
@@ -1229,14 +1279,33 @@ func (c *Controller) setStorageClusterDefaults(cluster *corev1.StorageCluster) e
 		logrus.Debugf("Failed to update driver: %v", err)
 	}
 
-	if toUpdate.Status.Phase == "" &&
-		toUpdate.Spec.CloudStorage != nil &&
+	// if no value is set for any of max_storage_nodes*, try to see if can set a default value
+	if toUpdate.Spec.CloudStorage != nil &&
+		toUpdate.Spec.CloudStorage.MaxStorageNodesPerZonePerNodeGroup == nil &&
 		toUpdate.Spec.CloudStorage.MaxStorageNodesPerZone == nil &&
 		toUpdate.Spec.CloudStorage.MaxStorageNodes == nil {
-		// Let's do this only when it's a fresh install of px
-		toUpdate.Spec.CloudStorage.MaxStorageNodesPerZone = new(uint32)
-		*toUpdate.Spec.CloudStorage.MaxStorageNodesPerZone = getDefaultMaxStorageNodesPerZone(zoneMap)
+		maxStorageNodesPerZone := uint32(0)
+		err = nil
+		if toUpdate.Status.Phase == "" {
+			// Let's do this only when it's a fresh install of px
+			maxStorageNodesPerZone = getDefaultMaxStorageNodesPerZone(zoneMap)
+		} else {
+			// Upgrade scenario
+			if c.isPxImageBeingUpdated(toUpdate) {
+				// If PX image is not changing, we don't have to update the values. This is to prevent pod restarts
+				maxStorageNodesPerZone, err = c.getCurrentMaxStorageNodesPerZone(cluster, nodeList, cloudProvider)
```

Comment:
The error received here will most likely never get printed. Whenever there is an error, we return a 0 value for maxStorageNodesPerZone.
We can probably just print the error here right away and change the below lines - 
```
if err == nil && maxStorageNodesPerZone != 0 {
   toUpdate.Spec.CloudStorage.MaxStorageNodesPerZone = &maxStorageNodesPerZone
   logrus.Infof("setting max_storage_nodes_per_zone=%v", maxStorageNodesPerZone)
}